components/brave_new_tab_ui/hooks/PrefHookManager.ts

```diff
@@ -0,0 +1,104 @@
+import { useCallback, useEffect, useState } from 'react'
+
+export type PrefListener<OnType, KeyType extends keyof OnType> = (name: KeyType, newValue: OnType[KeyType], oldValue?: OnType[KeyType]) => void
+
+/**
+ * A class to help wrapping mojom pref access in a React hook.
+ */
+export class PrefHookManager<T> {
+    #listeners: Map<keyof T, Array<PrefListener<T, keyof T>>> = new Map()
+    #lastValue?: T
+    #savePref: (key: keyof T, value: T[keyof T]) => void
+
+    /**
+     * 
+     * @param getInitialValue A function for triggering an initial get of the prefs this manager is responsible for. This function will be invoked when the |PrefHookManager| is constructed.
+     * @param savePref A function for saving a single pref and it's value.
+     * @param addListener A function allowing the |PrefHookManager| to subscribe to updates.
+     */
+    constructor(getInitialValue: () => Promise<T>, savePref: (key: keyof T, value: T[keyof T]) => void, addListener: (listener: (prefs: T) => void) => void) {
+        this.#savePref = savePref
+
+        addListener((prefs: T) => {
+            const oldPrefs = this.#lastValue ?? {} as Partial<T>
+            this.#lastValue = prefs
+
+            for (const [pref, listeners] of this.#listeners.entries()) {
+                const newValue = prefs[pref]
+                const oldValue = oldPrefs[pref]
+                if (newValue === oldValue) continue
+
+                for (const listener of listeners) { listener(pref, newValue, oldValue) }
+            }
+        })
```

Comment:
nit: perhaps a code comment about what this call to `addListener` is doing, e.g. `"Subscribe to changes for all values and, when any change, call listeners for each specific key"`

Reply:
:+1: good call